src/net/server/channel/handlers/ItemIdSortHandler.java

```diff
@@ -205,27 +212,39 @@ public final void handlePacket(SeekableLittleEndianAccessor slea, MapleClient c)
         ArrayList<Item> itemarray = new ArrayList<>();
         List<ModifyInventory> mods = new ArrayList<>();
         
-        for (short i = 1; i <= inventory.getSlotLimit(); i++) {
-            Item item = inventory.getItem(i);
-            if (item != null) {
-            	itemarray.add((Item) item.copy());
+
+        LinkedHashMap<Short, Item> recovery = inventory.cloneInventory();
+        for(Map.Entry<Short, Item> entry : recovery.entrySet()) {
+            if(entry.getValue() != null) {
+	        itemarray.add(entry.getValue());
+	    }
+        }
+
+        try {
+            int invTypeCriteria = (MapleInventoryType.getByType(inventoryType) == MapleInventoryType.EQUIP) ? PairedQuicksort.SORT_CRITERIA_LEVEL : PairedQuicksort.SORT_CRITERIA_QUANTITY;
+            int sortCriteria = (ServerConstants.USE_ITEM_SORT_BY_NAME) ? PairedQuicksort.SORT_CRITERIA_NAME : PairedQuicksort.SORT_CRITERIA_ITEM_ID;
+            PairedQuicksort pq = new PairedQuicksort(itemarray, sortCriteria, invTypeCriteria);
+            inventory.clearInventory();
+            
+            // Remove all items from inventory client action
+            for(Item item : itemarray) {
+                mods.add(new ModifyInventory(ModifyInventory.REMOVE_ITEM_OP, item));
+            }
+
+            // Remove all items from inventory client action
+            for(Item item : itemarray) {
+                mods.add(new ModifyInventory(ModifyInventory.ADD_ITEM_OP, item.copy())); //to prevent crashes
+                inventory.addItem(item);
```

Comment:
Item here should retain its position before the sort since it wasn't modified, so I'm not sure why the sort works. inventory.addItem(item) should be put before this line, but if the current code works then you can disregard.

Reply:
Good catch - swapping the two lines will make items update properly.